### DISCUSSION TITLE
TypeError: Expected a non-empty string while pushing data to a state

### DISCUSSION THREAD

**Discussion by chamodRash** [vercel/next.js#60355] [Help] [answered]
### Summary

**Problem Statement:**
I am encountering an issue while attempting to retrieve images from OpenAI's DALL-E API in my `ImagePage component.` Despite successfully obtaining image URLs in the `response.data` from the API, I am facing a `TypeError: Expected a non-empty string` when attempting to update the images state.

**Code Details:**
Below is the relevant section of the `ImagePage` component:

```
  const router = useRouter();
  const [images, setImages] = useState<string[]>([]);

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      prompt: "",
      amount: "1",
      resolution: "512x512",
    },
  });

  const isLoading = form.formState.isSubmitting;

  const onSubmit = async (values: z.infer<typeof formSchema>) => {
    try {
      setImages([]);

      const response = await axios.post("/api/image", values);
      console.log(`response.data`, response.data);

      const urls = response.data.map((image: { url: string }) => image.url);
      console.log(`urls`, urls);

      setImages(urls);
      console.log(`images`, images);

      form.reset();
    } catch (error: any) {
      // TODO: open Pro Modal
      console.log(`error`, error);
    } finally {
      router.refresh();
    }
  };
```
**Observations:**
1. Console Logs: The `response.data` and `urls` are correctly logged in the console, indicating that the image URLs are being received from the API.
2. State Update: The issue arises when updating the `images` state using `setImages(urls)`, resulting in an empty `images` state.

**Request for Assistance:**
I have been trying to debug this for two days. I want some guidance to understand why the `images state` remains empty despite receiving valid image URLs in the `response.data`.

### Additional information

```js
// This is how i display the images

<div className="grid grid-col-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-4 mt-8">
            {images.map((image) => (
              <Card key={image} className="rounded-lg overflow-hidden">
                <div className="relative aspect-square">
                  <Image alt="Image" fill src={image} />
                </div>
                <CardFooter className="p-2">
                  <Button
                    onClick={() => window.open(image)}
                    variant="secondary"
                    className="w-full">
                    <Download className="h-4 w-4 mr-2" />
                    Download
                  </Button>
                </CardFooter>
              </Card>
            ))}
          </div>
```

**Comment by don-esteban** [comment]
What you describe is a TypeScript error. It should appear during compilation and not at runtime. For my limited understanding, your code is a bit confusing to me and I do not clearly understand your problem description.

> I am getting the image url successfully but when i am trying to push that to a state it is returning TypeError: Expected a non-empty string.

Please show the error and surrounding code.

**Reply by chamodRash** [reply]
Thank you for your response. Sorry for the confusion, and I appreciate your willingness to help. I updated my question. Can you read it again answer? Thank you!

**Reply by don-esteban** [reply]
I still do not see the complete error message and the surrounding code. But anyhow. Did you try to narrow the problem? For example substitute

`const urls = response.data.map((image: { url: string }) => image.url);`

with

`const urls = ['some/valid/image/url.jpeg', 'some/valid/image/url.jpeg']`

What happens, if you do so?

If I remember correctly, state setters "schedule" state updates. The state itself reflects this in the next render phase. So

`console.log(`images`, images);`

will show the "old" state. The [new React docs](https://react.dev/reference/react/useState) explain that very well.

I think, you don't need `setImages([]);`.

Perhaps you have two independent errors in your code. A React and a TypeScript error.

**Comment by icyJoseph** [comment, marked as answer]
Hi @chamodRash 

Could you share your `next.config.js` file? More specifically, the `remotePatterns` section for images? I think you might have something incorrect there.

**Reply by icyJoseph** [reply]
@chamodRash as to why the `images` is apparently not updated, that's just how JS works:

```js
 setImages(urls);
 console.log(`images`, images);
```

When you do this, you need to understand two things. First, values captured in closures need to be mutated to see their change. However, you are setting `setImages` to a new array, coming from the mapping over the `response.data`, the array is `urls`. 

This is necessary because, React only re-renders, when the state changes, so we need to pass a new array. Mutating `images`, for example by pushing, and  doing `setImages(images)`, won't trigger a re-render.

Second, when you do `setImages(urls)`, you are telling React to update that state, and it'll do that and reflect the changes in the next render frame. React works by frames, not immediately. 

These are the two reasons, why it is only in the next render pass, that `images` will equal `urls`.  The images you are trying to log is captured in the closure, and you are setting the next frame's images value to `urls`, but it won't be visible within your callback, because that callback has closure over the current frame.

You might want to read here:

- https://dev.to/this-is-learning/the-cost-of-consistency-in-ui-frameworks-4agi

**Reply by chamodRash** [reply]
Thank you very much. I checked the next.congif.js file and there was a mistake. and it worked just fine.

**Reply by icyJoseph** [reply]
Awesome :D please don't forget to mark as answered!

**Reply by chamodRash** [reply]
Although it worked there is a new problem. Can you help me for that? 

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "oaidalleapiprodscus.blob.core.windows.net",
      },
    ],
  },
};

const withNextIntl = require("next-intl/plugin")(
  // This is the default (also the `src` folder is supported out of the box)
  "./i18n.ts"
);

// module.exports = withNextIntl({
//   nextConfig,
//   api: {
//     bodyParser: true,
//   },
// });
module.exports = nextConfig;
```

in this commented next-intl is the one i should use in order to use the app with multi languages. but if i export nextConfig inside this withNextIntl it won't work. So what's the solution for this?
